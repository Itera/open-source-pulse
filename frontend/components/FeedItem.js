// @flow
import React from 'react';
import moment from 'moment';
import { get, find } from 'lodash/fp';
import styled from 'styled-components';

import FeedEntryTypes from '../FeedEntryTypes';
import type { User } from '../../types/User';

type Props = {
  type: string,
  url: string,
  user: User,
  timestamp: string,
};

function titleFromUrl(url) {
  const parts = url.match(/com\/([^/]+)\/([^/]+)(?:\/(?:pull|issues)\/(\d+))?/) || [];
  return parts.slice(1, 3).join(' / ') + (parts[3] ? ` #${parts[3]}` : '');
}

const Date = styled.em`
  color: gray;
  font-size: 0.8em;
`;

const InfoContainer = styled.span`
  display: flex;
  flex-direction: column;
  text-align: left;
`;

const ProfilePicture = styled.img`
  width: 100px;
  height: 100px;
  margin-right: 1rem;
`;

const RepoLink = styled.a`
  margin-top: 0.5rem;
  font-size: 1.5em;
  color: black;
`;

const ProfileLink = styled.a`
  color: black;
  text-decoration: none;
`;

function FeedItem({ user, type, url, timestamp }: Props) {
  const profileUrl = `https://github.com/${user.username}`;
  return (
    <div className="FeedItem">
      {user &&
        <div className="flex flex--center">
          <ProfileLink href={profileUrl}>
            <ProfilePicture src={user.photos[0]} alt={user.displayName} />
          </ProfileLink>
          <InfoContainer>
            <Date>{moment(new Date(timestamp)).format('DD.MM.YYYY')}</Date>
            <span>
              <ProfileLink href={profileUrl}><strong>{user.displayName}</strong></ProfileLink>
              &nbsp;
              {get('feedText')(find({ key: type })(FeedEntryTypes))}
            </span>
            <RepoLink href={url}>{titleFromUrl(url)}</RepoLink>
          </InfoContainer>
        </div>
      }
    </div>
  );
}

export default FeedItem;
