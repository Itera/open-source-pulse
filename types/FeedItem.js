export type FeedItem = {|
  timestamp: string,
  type: string,
  user: User,
  url: string
|};

export type FeedItemInput = {|
  type: string,
  username: string,
  url: string
|};
