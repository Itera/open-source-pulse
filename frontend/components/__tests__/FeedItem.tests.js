// @flow
/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import FeedItem from '../FeedItem';

describe('FeedItem', () => {
  const user = { username: 'albus', displayName: 'Albus Dumbledore', photos: ['albus.png'] };

  it('should render correctly for type CREATED_PR', () => {
    const wrapper = shallow(
      <FeedItem
        timestamp={new Date(2016, 4, 4, 10, 0).toISOString()}
        url="https://github.com/flowtype/flow-typed/pull/406"
        type="CREATED_PR"
        user={user}
      />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly for type WORKING_ON_ISSUE', () => {
    const wrapper = shallow(
      <FeedItem
        timestamp={new Date(2016, 4, 4, 10, 0).toISOString()}
        url="https://github.com/Itera/open-source-pulse/issues/16"
        type="WORKING_ON_ISSUE"
        user={user}
      />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly for type CODING_REPOSITORIY', () => {
    const wrapper = shallow(
      <FeedItem
        timestamp={new Date(2016, 4, 4, 10, 0).toISOString()}
        url="https://github.com/flowtype/flow-typed"
        type="CODING_REPOSITORY"
        user={user}
      />
    );

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
