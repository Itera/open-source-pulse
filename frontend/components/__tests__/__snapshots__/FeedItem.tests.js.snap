exports[`FeedItem should render correctly for type CODING_REPOSITORIY 1`] = `
<div
  className="FeedItem">
  <div
    className="flex flex--center">
    <styled.a
      href="https://github.com/albus">
      <styled.img
        alt="Albus Dumbledore"
        src="albus.png" />
    </styled.a>
    <styled.span>
      <styled.em>
        01.11.2016
      </styled.em>
      <span>
        <styled.a
          href="https://github.com/albus">
          <strong>
            Albus Dumbledore
          </strong>
        </styled.a>
         
        is working on
      </span>
      <styled.a
        href="https://github.com/flowtype/flow-typed">
        flowtype / flow-typed
      </styled.a>
    </styled.span>
  </div>
</div>
`;

exports[`FeedItem should render correctly for type CREATED_PR 1`] = `
<div
  className="FeedItem">
  <div
    className="flex flex--center">
    <styled.a
      href="https://github.com/albus">
      <styled.img
        alt="Albus Dumbledore"
        src="albus.png" />
    </styled.a>
    <styled.span>
      <styled.em>
        01.11.2016
      </styled.em>
      <span>
        <styled.a
          href="https://github.com/albus">
          <strong>
            Albus Dumbledore
          </strong>
        </styled.a>
         
        created a PR on
      </span>
      <styled.a
        href="https://github.com/flowtype/flow-typed/pull/406">
        flowtype / flow-typed #406
      </styled.a>
    </styled.span>
  </div>
</div>
`;

exports[`FeedItem should render correctly for type WORKING_ON_ISSUE 1`] = `
<div
  className="FeedItem">
  <div
    className="flex flex--center">
    <styled.a
      href="https://github.com/albus">
      <styled.img
        alt="Albus Dumbledore"
        src="albus.png" />
    </styled.a>
    <styled.span>
      <styled.em>
        01.11.2016
      </styled.em>
      <span>
        <styled.a
          href="https://github.com/albus">
          <strong>
            Albus Dumbledore
          </strong>
        </styled.a>
         
        is working on
      </span>
      <styled.a
        href="https://github.com/Itera/open-source-pulse/issues/16">
        Itera / open-source-pulse #16
      </styled.a>
    </styled.span>
  </div>
</div>
`;
